components/tidb_query_vec_executors/src/index_scan_executor.rs

```diff
@@ -232,59 +321,147 @@ impl IndexScanExecutorImpl {
+        )?;
+
         Ok(())
     }
 
-    fn process_kv_pair_old(
+    fn process_normal_new_collation_value(
+        &mut self,
+        mut key: &[u8],
+        value: &[u8],
+        columns: &mut LazyBatchColumnVec,
+    ) -> Result<()> {
+        use DecodeHandleStrategy::*;
+        let tail_len = value[0] as usize;
+
+        if tail_len > value.len() {
+            return Err(other_err!("`tail_len`: {} is corrupted", tail_len));
+        }
+
+        let restore_values = &value[1..value.len() - tail_len];
+        self.extract_columns_from_row_format(restore_values, columns)?;
+
+        match self.decode_handle_strategy {
+            NoDecode => {}
+            DecodeIntHandle if tail_len < 8 => {
+                key = &key[table::PREFIX_LEN + table::ID_LEN..];
+                datum::skip_n(&mut key, self.columns_id_without_handle.len())?;
+                let handle = self.decode_handle_from_key(key)?;
+                columns[self.columns_id_without_handle.len()]
+                    .mut_decoded()
+                    .push_int(Some(handle));
+            }
+            DecodeIntHandle => {
+                let handle = self.decode_handle_from_value(&value[value.len() - tail_len..])?;
+                columns[self.columns_id_without_handle.len()]
+                    .mut_decoded()
+                    .push_int(Some(handle));
+            }
+            DecodeCommonHandle => {
+                key = &key[table::PREFIX_LEN + table::ID_LEN..];
+                datum::skip_n(&mut key, self.columns_id_without_handle.len())?;
+                Self::extract_columns_from_datum_format(
+                    &mut key,
+                    &mut columns[self.columns_id_without_handle.len()..],
+                )?;
+            }
+        }
+        Ok(())
+    }
+
+    fn process_normal_old_collation_value(
         &mut self,
         key: &[u8],
         value: &[u8],
         columns: &mut LazyBatchColumnVec,
     ) -> Result<()> {
-        check_index_key(key)?;
+        use DecodeHandleStrategy::*;
+
         // The payload part of the key
         let mut key_payload = &key[table::PREFIX_LEN + table::ID_LEN..];
 
-        for i in 0..self.columns_id_without_handle.len() {
-            let (val, remaining) = datum::split_datum(key_payload, false)?;
-            columns[i].mut_raw().push(val);
-            key_payload = remaining;
-        }
+        Self::extract_columns_from_datum_format(
+            &mut key_payload,
+            &mut columns[..self.columns_id_without_handle.len()],
+        )?;
 
-        if self.decode_handle {
+        match self.decode_handle_strategy {
+            NoDecode => {}
             // For normal index, it is placed at the end and any columns prior to it are
             // ensured to be interested. For unique index, it is placed in the value.
-            let handle_val = if key_payload.is_empty() {
+            DecodeIntHandle if key_payload.is_empty() => {
```

Comment:
This function is called `process_normal_xxx` but this match arm handles unique index?

Reply:
Yes, we should rename it to `process_old_collation_value`